From e7ee85880ce2779180bf7fbc3cf199ada495484b Mon Sep 17 00:00:00 2001
From: Miller Puckette <msp@ucsd.edu>
Date: Tue, 24 Aug 2021 22:02:38 -0700
Subject: [PATCH 1/4] add m_local.h include if PD not defined

---
 src/m_pd.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/m_pd.h b/src/m_pd.h
index 59841935..edc5db30 100644
--- a/src/m_pd.h
+++ b/src/m_pd.h
@@ -6,6 +6,11 @@
 
 #if defined(_LANGUAGE_C_PLUS_PLUS) || defined(__cplusplus)
 extern "C" {
+#endif
+
+  /* if you can't get the build system to add defs, you can do it this way */
+#ifndef PD
+#include "m_local.h"
 #endif
 
 #define PD_MAJOR_VERSION 0
-- 
2.33.1

