From ca00daa0c5e51ab263a4c7bf239197bdb97e3a98 Mon Sep 17 00:00:00 2001
From: Miller Puckette <msp@ucsd.edu>
Date: Sat, 11 Dec 2021 09:48:29 -0800
Subject: [PATCH 4/4] fix espd-specific compile warning that generates an
 error?

---
 src/x_text.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/x_text.c b/src/x_text.c
index 9bc91247..ce8e348c 100644
--- a/src/x_text.c
+++ b/src/x_text.c
@@ -98,7 +98,7 @@ static void textbuf_open(t_textbuf *x)
             x, 600, 340, x->b_sym->s_name,
                  sys_hostfontsize(glist_getfont(x->b_canvas),
                     glist_getzoom(x->b_canvas)));
-        sprintf(buf, ".x%lx", x);
+        sprintf(buf, ".x%lx", (unsigned long)x);
         x->b_guiconnect = guiconnect_new(&x->b_ob.ob_pd, gensym(buf));
         textbuf_senditup(x);
     }
-- 
2.33.1

