diff --git a/src/d_fft.c b/src/d_fft.c
index 6ba31144..47a28d60 100644
--- a/src/d_fft.c
+++ b/src/d_fft.c
@@ -118,7 +118,7 @@ static void sigfft_dspx(t_sigfft *x, t_signal **sp, t_int *(*f)(t_int *w))
         dsp_add_zero(sp[3]->s_vec, length * nchans);
         return;
     }
-    for (ch = 0; ch < nchans; ch++)
+    for (ch = nchans; ch--; )
     {
         t_sample *in1 = sp[0]->s_vec + ch * length;
         t_sample *in2 = sp[1]->s_vec + ch * length;
@@ -128,13 +128,13 @@ static void sigfft_dspx(t_sigfft *x, t_signal **sp, t_int *(*f)(t_int *w))
             dsp_add(sigfft_swap, 3, out1, out2, (t_int)length);
         else if (out1 == in2)
         {
-            dsp_add(copy_perform, 3, in2, out2, (t_int)length);
-            dsp_add(copy_perform, 3, in1, out1, (t_int)length);
+            dsp_add_copy(in2, out2, length);
+            dsp_add_copy(in1, out1, length);
         }
         else
         {
-            if (out1 != in1) dsp_add(copy_perform, 3, in1, out1, (t_int)length);
-            if (out2 != in2) dsp_add(copy_perform, 3, in2, out2, (t_int)length);
+            if (out1 != in1) dsp_add_copy(in1, out1, length);
+            if (out2 != in2) dsp_add_copy(in2, out2, length);
         }
         dsp_add(f, 3, out1, out2, (t_int)length);
     }
@@ -213,13 +213,13 @@ static void sigrfft_dsp(t_sigrfft *x, t_signal **sp)
         dsp_add_zero(sp[2]->s_vec, length * nchans);
         return;
     }
-    for (ch = 0; ch < nchans; ch++)
+    for (ch = nchans; ch--; )
     {
         t_sample *in1 = sp[0]->s_vec + ch * length;
         t_sample *out1 = sp[1]->s_vec + ch * length;
         t_sample *out2 = sp[2]->s_vec + ch * length;
         if (in1 != out1)
-            dsp_add(copy_perform, 3, in1, out1, (t_int)length);
+            dsp_add_copy(in1, out1, length);
         dsp_add(sigrfft_perform, 2, out1, (t_int)length);
         dsp_add(sigrfft_flip, 3, out1 + (n2+1), out2 + n2, (t_int)(n2-1));
         dsp_add_zero(out1 + (n2+1), ((n2-1)&(~7)));
@@ -286,7 +286,7 @@ static void sigrifft_dsp(t_sigrifft *x, t_signal **sp)
         dsp_add_zero(sp[2]->s_vec, length * nchans);
         return;
     }
-    for (ch = 0; ch < nchans; ch++)
+    for (ch = nchans; ch--; )
     {
         t_sample *in1 = sp[0]->s_vec + ch * length;
         t_sample *in2 = sp[1]->s_vec + ch * length;
@@ -294,11 +294,11 @@ static void sigrifft_dsp(t_sigrifft *x, t_signal **sp)
         if (in2 == out1)
         {
             dsp_add(sigrfft_flip, 3, out1+1, out1 + length, (t_int)(n2-1));
-            dsp_add(copy_perform, 3, in1, out1, (t_int)(n2+1));
+            dsp_add_copy(in1, out1, n2+1);
         }
         else
         {
-            if (in1 != out1) dsp_add(copy_perform, 3, in1, out1, (t_int)(n2+1));
+            if (in1 != out1) dsp_add_copy(in1, out1, n2+1);
             dsp_add(sigrfft_flip, 3, in2+1, out1 + length, (t_int)(n2-1));
         }
         dsp_add(sigrifft_perform, 2, out1, (t_int)length);
diff --git a/src/g_io.c b/src/g_io.c
index ae42e038..427937e3 100644
--- a/src/g_io.c
+++ b/src/g_io.c
@@ -16,6 +16,12 @@ life elsewhere. */
 #include "m_pd.h"
 #include "g_canvas.h"
 #include <string.h>
+#ifdef ESP_PLATFORM
+#include "esp_attr.h"
+#endif
+#ifndef IRAM_ATTR
+#define IRAM_ATTR
+#endif
 
 static int symbol2resamplemethod(t_symbol*s)
 {
@@ -123,7 +129,7 @@ static void vinlet_free(t_vinlet *x)
     int i;
     canvas_rminlet(x->x_canvas, x->x_inlet);
     if (x->x_rb)
-        for (i = 0; i < x->x_nchans; i++)
+        for (i = x->x_nchans; i--; )
     {
         freebytes((x->x_rb)[i].r_buf, x->x_buflength * sizeof(t_sample));
         resample_free(&(x->x_rb)[i].r_updown);
@@ -142,7 +148,7 @@ int vinlet_issignal(t_vinlet *x)
     return (x->x_rb != 0);
 }
 
-t_int *vinlet_perform(t_int *w)
+IRAM_ATTR t_int *vinlet_perform(t_int *w)
 {
     t_vinlet *x = (t_vinlet *)(w[1]);
     t_sample *out = (t_sample *)(w[2]);
@@ -195,7 +201,7 @@ static void vinlet_dsp(t_vinlet *x, t_signal **sp)
 }
 
     /* prolog code: loads buffer from parent patch */
-t_int *vinlet_doprolog(t_int *w)
+IRAM_ATTR t_int *vinlet_doprolog(t_int *w)
 {
     t_vinlet *x = (t_vinlet *)(w[1]);
     t_sample *in = (t_sample *)(w[2]), *out;
@@ -260,7 +266,7 @@ void vinlet_dspprolog(struct _vinlet *x, t_signal **parentsigs,
             bufsize = myvecsize;
         if (bufsize != (oldbufsize = x->x_buflength))
         {
-            for (i = 0; i < x->x_nchans; i++)
+            for (i = x->x_nchans; i--; )
             {
                 x->x_rb[i].r_buf = (t_sample *)t_resizebytes(x->x_rb[i].r_buf,
                     oldbufsize * sizeof(t_sample), bufsize * sizeof(t_sample));
@@ -298,7 +304,7 @@ void vinlet_dspprolog(struct _vinlet *x, t_signal **parentsigs,
                 }
             }
         }
-        else for (i = 0; i < x->x_nchans; i++)
+        else for (i = x->x_nchans; i--; )
             memset((char *)(x->x_rb[i].r_buf), 0, bufsize * sizeof(t_sample));
         x->x_directsignal = 0;
     }
@@ -433,7 +439,7 @@ static void voutlet_free(t_voutlet *x)
     int i;
     canvas_rmoutlet(x->x_canvas, x->x_parentoutlet);
     if (x->x_rb)
-        for (i = 0; i < x->x_nchans; i++)
+        for (i = x->x_nchans; i--; )
     {
         freebytes((x->x_rb)[i].r_buf, x->x_buflength * sizeof(t_sample));
         resample_free(&(x->x_rb)[i].r_updown);
@@ -454,7 +460,7 @@ int voutlet_issignal(t_voutlet *x)
 }
 
     /* LATER optimize for non-overlapped case where the "+=" isn't needed */
-t_int *voutlet_perform(t_int *w)
+IRAM_ATTR t_int *voutlet_perform(t_int *w)
 {
     t_voutlet *x = (t_voutlet *)(w[1]);
     t_sample *in = (t_sample *)(w[2]), *buf= (t_sample *)(w[3]);
@@ -477,7 +483,7 @@ t_int *voutlet_perform(t_int *w)
 }
 
     /* epilog code for blocking: write buffer to parent patch */
-static t_int *voutlet_doepilog(t_int *w)
+static IRAM_ATTR t_int *voutlet_doepilog(t_int *w)
 {
     t_voutlet *x = (t_voutlet *)(w[1]);
     t_sample *in, *out = (t_sample *)(w[2]), *buf = (t_sample *)(w[3]);
@@ -495,7 +501,7 @@ static t_int *voutlet_doepilog(t_int *w)
     /* It's dumb that we have to offer a duplicate but we don't have the
     goddam resampling buffer allocated by the time this goes on the DSP
     chain so we have to look it up dynamically. */
-static t_int *voutlet_doepilog_resample(t_int *w)
+static IRAM_ATTR t_int *voutlet_doepilog_resample(t_int *w)
 {
     t_voutlet *x = (t_voutlet *)(w[1]);
     t_sample *in, *out = ((t_resample *)(w[2]))->s_vec,
@@ -555,7 +561,7 @@ void voutlet_dspprolog(struct _voutlet *x, t_signal **parentsigs,
             buflength = myvecsize;
         if (buflength != x->x_buflength)
         {
-            for (i = 0; i < x->x_nchans; i++)
+            for (i = x->x_nchans; i--; )
             {
                 x->x_rb[i].r_buf = (t_sample *)t_resizebytes(x->x_rb[i].r_buf,
                     x->x_buflength * sizeof(t_sample),
